hail/src/main/scala/is/hail/linalg/LinalgCodeUtils.scala

```diff
@@ -32,41 +33,19 @@ object LinalgCodeUtils {
 
   def createColumnMajorCode(pndv: PNDArrayValue, cb: EmitCodeBuilder, region: Value[Region]): PNDArrayCode = {
     val shape = pndv.shapes(cb)
-    val shapeStruct = pndv.pt.makeShapeStruct(shape, region, cb)
-    val stridesStruct = pndv.pt.makeColumnMajorStridesStruct(shape, region, cb)
-    val dataLength = pndv.pt.numElements(shape)
+    val pt = pndv.pt.asInstanceOf[PCanonicalNDArray]
+    val strides = pt.makeColumnMajorStrides(shape, region, cb)
+    val dataLength = cb.newLocal[Int]("nda_create_column_major_len", pt.numElements(shape).toI)
+    val dataType = pt.dataType
 
-    val outputElementPType = pndv.pt.elementType
-    val idxVars = Array.tabulate(pndv.pt.nDims) { _ => cb.emb.genFieldThisRef[Long]() }.toFastIndexedSeq
-
-    def loadElement(ndValue: PNDArrayValue) = {
-      ndValue.pt.loadElementToIRIntermediate(idxVars, ndValue.value.asInstanceOf[Value[Long]], cb)
-    }
-
-    val srvb = new StagedRegionValueBuilder(cb.emb, pndv.pt.data.pType, region)
-
-    val body =
-      Code(
-        srvb.addIRIntermediate(outputElementPType)(loadElement(pndv)),
-        srvb.advance()
-      )
-
-    val columnMajorLoops = idxVars.zipWithIndex.foldLeft(body) { case (innerLoops, (dimVar, dimIdx)) =>
-      Code(
-        dimVar := 0L,
-        Code.whileLoop(dimVar < shape(dimIdx),
-          innerLoops,
-          dimVar := dimVar + 1L
-        )
-      )
+    val (addElem, finish) = dataType.constructFromFunctions(cb, region, dataLength, deepCopy = false)
+    val idx = cb.newLocal[Int]("nda_create_column_major_idx", 0)
+    SNDArray.forEachIndex(cb, shape, "nda_create_column_major") { case (cb, idxVars) =>
+      addElem(cb, idx, IEmitCode.present(cb, pndv.loadElement(idxVars, cb).asPCode))
```

Comment:
You're getting assertion errors from here. I think it's not ok to `IEmitCode.present` multiple times inside of an `IEmitCode.map`, which is what you're indirectly doing here within the NDArrayEmitter

Reply:
debugged that yesterday -- it's actually that IEmitCode.present is totally busted. I pulled out the fix in #9962

Reply:
(assigned to you)

Reply:
I'm now getting failures about mismatched shapes in BLockMatrixDot examples (and others), I must some incorrect shape calculations.